From bcd52fed506d08f713bd04889638a34feb66dad1 Mon Sep 17 00:00:00 2001
From: Todd Lipcon <todd@apache.org>
Date: Fri, 13 Jul 2012 05:00:47 +0000
Subject: [PATCH 1308/1344] HDFS-3652. FSEditLog failure removes the wrong edit stream when storage dirs have same name. Contributed by Todd Lipcon.

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1361047 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit b6c13375c54dc98666bd8cd6a49012536f07973d)

When the NameNode is configured with multiple storage directories, and
those storage directories have the same terminal path component (eg
/data/1/nn and /data/2/nn), some rare failure scenarios may result in a
corrupt namesystem.

Reason: Bug
Author: Todd Lipcon
Ref: CDH-6757
---
 .../hadoop/hdfs/server/namenode/FSEditLog.java     |    4 +++-
 .../namenode/TestStorageDirectoryFailure.java      |    8 +++++---
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/hdfs/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java b/src/hdfs/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java
index a9b2f24..234ae26 100644
--- a/src/hdfs/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java
+++ b/src/hdfs/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java
@@ -510,8 +510,10 @@ public class FSEditLog {
                                        .get(idx)).getFile()
                                        .getParentFile().getParentFile();
       File parentDir = getStorageDirForStream(idx);
-      if (parentDir.getName().equals(sd.getRoot().getName())) {
+      if (parentDir.getAbsolutePath().equals(
+            sd.getRoot().getAbsolutePath())) {
         editStreams.remove(idx);
+        idx--;
       }
     }
     exitIfNoStreams();
diff --git a/src/test/org/apache/hadoop/hdfs/server/namenode/TestStorageDirectoryFailure.java b/src/test/org/apache/hadoop/hdfs/server/namenode/TestStorageDirectoryFailure.java
index 505005a..3c5b668 100644
--- a/src/test/org/apache/hadoop/hdfs/server/namenode/TestStorageDirectoryFailure.java
+++ b/src/test/org/apache/hadoop/hdfs/server/namenode/TestStorageDirectoryFailure.java
@@ -63,9 +63,11 @@ public class TestStorageDirectoryFailure {
     String baseDir = System.getProperty("test.build.data", "/tmp");
     File dfsDir = new File(baseDir, "dfs");
     nameDirs = new ArrayList<String>();
-    nameDirs.add(new File(dfsDir, "name1").getPath());
-    nameDirs.add(new File(dfsDir, "name2").getPath());
-    nameDirs.add(new File(dfsDir, "name3").getPath());
+    // Have all the name dirs with the same filename: important for regression
+    // testing HDFS-3652.
+    nameDirs.add(new File(new File(dfsDir, "name1"), "nn").getPath());
+    nameDirs.add(new File(new File(dfsDir, "name2"), "nn").getPath());
+    nameDirs.add(new File(new File(dfsDir, "name3"), "nn").getPath());
 
     conf.set("dfs.name.dir", StringUtils.join(nameDirs, ","));
     conf.set("dfs.data.dir", new File(dfsDir, "data").getPath());
-- 
1.7.0.4

